### DISCUSSION TITLE
Custom Rule Referencing Another Table

### DISCUSSION THREAD

**Discussion by WhitestRabbit** [adonisjs/core#2870] [Help] [answered]
Let's say that I am writing a custom validation rule for table A. Would it be possible to reference the fields of another table B from within that custom rule? I tried making a database query, but the asynchronous nature broke the rest of the validation.

**Comment by thetutlage** [comment]
Can you share what you have written so far to implement your rule?

**Reply by WhitestRabbit** [reply]
I am creating a resource "shop" that can be connected to some "options" with a many-to-many relationship through an pivot table. The body of the request I am sending contains some fields to first create the shop and I am also sending an options field which contains an object that looks like this in the JSON:

`"options": { "2": "", "3": {"value": "test value 6"}, "4": {"value": "test value 7"}}`

The request being validated is passed through this schema:

`public schema = schema.create({
		name: schema.string({ trim: true }),
		...other fields,
		options: schema.object([rules.option_format()]).anyMembers()
	})`

And the custom rule 'options_format' for the options about to be attached is a little like this:
`validator.rule('option_format', (value, _, { pointer, arrayExpressionPointer, errorReporter }) => {
    for (const id in value) {
        if(/* option number does not exist in the options table */) {
            errorReporter.report(pointer, 'option_format', `Option id ${id} does not exist in the options table`, arrayExpressionPointer);
        }

        if (isNaN(+id)) {
            errorReporter.report(pointer, 'option_format', `Option id ${id} is not a number`, arrayExpressionPointer);
        }

        if(value[id] === "") {
            continue;
        }

        if(typeof(value[id]) !== "object") {
            errorReporter.report(pointer, 'option_format', `Option should either be an empty string or an object with a value, you tried inserting: ${value[id]}`, arrayExpressionPointer);
        } 
...more validation on the options`

**Comment by arthurcvm** [comment, marked as answer]
Hi @WhitestRabbit, if your problem is in async call it's very simple tho solve, look the model of custom validator with async below:

```ts
validator.rule('nameOfRule', async (value, _, { pointer, arrayExpressionPointer, errorReporter }) => {
  if (isNaN(value)) {
    return
  }

  const permission = await Permission.query().where('id', value).first()

  if (!permission) {
    return
  }

  if (permission.type !== 'page') {
    errorReporter.report(pointer, 'nameOfRule', 'Invalid permission type', arrayExpressionPointer)
  }
},
() => {
  return {
    async: true
  }
})
```

In the end we pass the `compileFn` returning the option `async: true`.

**Reply by arthurcvm** [reply]
@thetutlage pls accept [my PR](https://github.com/adonisjs/docs.adonisjs.com/pull/21) in docs with this tip

**Reply by WhitestRabbit** [reply]
It works! Thank you!
